File: src/lib/video.ts
// Video provider adapter seam.
//
// The backend mints the join credential (Agora app id / channel / token / uid).
// The real Agora Web SDK is NOT wired yet — this stub is the single seam where it
// will plug in. The frontend never generates tokens; it only consumes the
// server-minted credential.
import type { VideoJoin } from "@/api/types";

export interface VideoSession {
  channel: string;
  disconnect(): void;
}

export interface VideoProvider {
  connect(join: VideoJoin): Promise<VideoSession>;
}

/** Placeholder provider — logs the credential and returns a no-op session. */
export const stubVideoProvider: VideoProvider = {
  async connect(join: VideoJoin): Promise<VideoSession> {
    // Real implementation (later phase):
    //   const client = AgoraRTC.createClient({ mode: "rtc", codec: "vp8" });
    //   await client.join(join.agoraAppId, join.channel, join.agoraToken, join.uid);
    if (import.meta.env.DEV) {
      // eslint-disable-next-line no-console
      console.info("[video] (stub) joining channel", join.channel, "as uid", join.uid);
    }
    return {
      channel: join.channel,
      disconnect() {
        /* real impl: client.leave() */
      },
    };
  },
};

File: backend/infrastructure/gateways/video.py
"""
Video provider — STUB ONLY.

This is the seam where Agora will plug in later. It intentionally does NOT call
Agora and does NOT mint real tokens. The real adapter (AgoraVideoProvider) will
implement the same VideoProvider port and generate genuine RTC tokens
server-side. Use cases depend on the port, never on this class directly.
"""
from datetime import timedelta

from django.utils import timezone

from application.ports.gateways import VideoProvider, VideoToken


class StubVideoProvider(VideoProvider):
    provider_name = "stub"

    def create_channel(self, *, session_id) -> str:
        return f"session-{session_id}"

    def issue_join(self, *, channel, identity) -> VideoToken:
        # Deterministic, obviously-fake credential. NOT an Agora token.
        return VideoToken(
            provider=self.provider_name,
            channel=channel,
            token=f"stub-token::{channel}::{identity}",
            uid=str(identity),
            expires_at=timezone.now() + timedelta(hours=1),
            app_id="stub-app-id",  # real AgoraVideoProvider supplies the genuine app id
        )
